README.md
# tcp-ping

[![checks](https://github.com/justintaddei/tcp-ping/actions/workflows/tests.yml/badge.svg?branch=master)](https://github.com/justintaddei/tcp-ping/actions/workflows/tests.yml)
![](https://img.shields.io/github/issues-raw/justintaddei/tcp-ping.svg?style=flat)
![](https://img.shields.io/npm/v/@network-utils/tcp-ping.svg?style=flat)
![](https://img.shields.io/npm/dt/@network-utils/tcp-ping.svg?style=flat)
![](https://img.shields.io/badge/language-typescript-blue.svg?style=flat)
![](https://img.shields.io/npm/l/@network-utils/tcp-ping.svg?style=flat)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat)](https://github.com/prettier/prettier)

A simple promise-based TCP ping util, written in Typescript, to test the reachability and latency of a host.

## Installation

```bash
$ npm install @network-utils/tcp-ping --save
```

## Usage

### `ping(options?: Partial<IPingOptions>, progress?: (progress, total) => void): Promise<IPingResult>`

Pings the given host and returns an object containing the latency of the connection
and any errors that may have occured.

> **NOTE:** Attempts are not concurrent. As such if a host is unreachable and you provide options { attempts: 60, timeout: 1000 } then `ping` will not resolve for a full minute!

```typescript
import { ping } from '@network/tcp-ping'

ping({
  // The IP address or hostname of the host you want
  // to ping. Defaults to 127.0.0.1 (localhost)
  address: '192.168.1.47',
  // How many times do you want want to
  // attempt to reach the host? Default is 10
  attempts: 10,
  // What port do you want to connect on?
  // Default is 80
  port: 80,
  // How long do you want to wait (in milliseconds)
  // before assuming an attempt has failed?
  // Default is 3000 (3 seconds)
  timeout: 3000
}, update).then(result => {
  console.log('ping result:', result)


    // ping result:
    {
      averageLatency: 19.2753,
      errors: [
        {
          // Which attempt failed
          attempt: 3,
          error: Error('Request timeout')
        }
      ],
      maximumLatency: 35.1978,
      minimumLatency: 3.7716,
      options: {
        address: '192.168.1.47',
        attempts: 10,
        port: 80,
        timeout: 3000
      }
    }
})

function update(progress, total) {
  console.log(progress, '/', total)
  /*
    1 / 10
    2 / 10
    3 / 10
    ...
  */
}
```

### `probe(port: number, address?: string, timeout?: number): Promise<boolean>`

Makes one attempt to reach the host and returns a `boolean` indicating whether or not it was successful.  
If `address` is not provided it will default to `'127.0.0.1'`.  
If `timeout` is not provided it will default to `3000`.

```typescript
import { probe } from '@network/tcp-ping'

probe(80, '192.168.1.47', 500).then(hostReachable => {
  if (hostReachable) console.log('The host is reachable 🙌')
  else console.log('The host is not reachable 🤐')
})

// Or

const hostReachable = await probe(80, '192.168.1.47', 500)
```

## Errors

- All methods will throw a `"Negative port"` error if `port < 1`.

---

## Testing

```bash
$ git clone https://github.com/justintaddei/tcp-ping.git
$ cd tcp-ping
$ npm install
$ npm test
```

## License

MIT


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.2.3](https://github.com/justintaddei/tcp-ping/compare/v1.2.2...v1.2.3) (2023-05-09)

### [1.2.2](https://github.com/justintaddei/tcp-ping/compare/v1.2.1...v1.2.2) (2023-05-09)


### Bug Fixes

* **deps:** bump actions/cache from 2.1.6 to 2.1.7 ([349eaad](https://github.com/justintaddei/tcp-ping/commit/349eaad59694eb69c4bfac742a2dadb58e009856))
* **deps:** bump actions/cache from 2.1.7 to 3.0.1 ([827b0d1](https://github.com/justintaddei/tcp-ping/commit/827b0d16fccb6365bb1ec48ec92986247e071da5))
* **deps:** bump actions/cache from 3.0.1 to 3.0.2 ([848168a](https://github.com/justintaddei/tcp-ping/commit/848168a389e0b1fb66d39fe652c1c898a57dca36))
* **deps:** bump actions/cache from 3.0.2 to 3.0.3 ([5f7bac5](https://github.com/justintaddei/tcp-ping/commit/5f7bac581a7aa029ae7b7b0838ce872e5bc99abd))
* **deps:** bump actions/cache from 3.0.3 to 3.0.4 ([a0f058b](https://github.com/justintaddei/tcp-ping/commit/a0f058b1ae5a474f914796ef7510833a8c899707))
* **deps:** bump actions/cache from 3.0.4 to 3.3.1 ([afb6146](https://github.com/justintaddei/tcp-ping/commit/afb61460c921209d41259876a32bff114ab7a289))
* **deps:** bump actions/setup-node from 2.2.0 to 2.3.0 ([0caab2c](https://github.com/justintaddei/tcp-ping/commit/0caab2cb2a5c0a3ba45c3e9edd707924bd3f5139))
* **deps:** bump actions/setup-node from 2.3.0 to 2.4.0 ([7c5baab](https://github.com/justintaddei/tcp-ping/commit/7c5baabcb008de0ccae2cec117611966945a1227))
* **deps:** bump actions/setup-node from 2.4.0 to 2.4.1 ([0c2e3da](https://github.com/justintaddei/tcp-ping/commit/0c2e3da412d8c1c2ef5b652de1cc7b98c2893723))
* **deps:** bump actions/setup-node from 2.4.1 to 2.5.1 ([891e0ba](https://github.com/justintaddei/tcp-ping/commit/891e0ba623b41c47a2820443cb032b68a4c29dd6))
* **deps:** bump actions/setup-node from 2.5.1 to 3 ([7efe91f](https://github.com/justintaddei/tcp-ping/commit/7efe91f274f775028cc94deab8dbaa2997cea06a))
* **deps:** bump ajv from 6.10.2 to 6.12.6 ([0cb40cb](https://github.com/justintaddei/tcp-ping/commit/0cb40cb67934b5c92117ba32085e62a3533f42ec))
* **deps:** bump dependabot/fetch-metadata from 1.1.1 to 1.2.0 ([9d7168a](https://github.com/justintaddei/tcp-ping/commit/9d7168ace09eeac2e19803c932a17b2f4ed7e66a))
* **deps:** bump dependabot/fetch-metadata from 1.2.0 to 1.2.1 ([9d3de02](https://github.com/justintaddei/tcp-ping/commit/9d3de0282c878fe51a6c57c4ee4d0e574cf28dfa))
* **deps:** bump dependabot/fetch-metadata from 1.2.1 to 1.3.0 ([0757cf3](https://github.com/justintaddei/tcp-ping/commit/0757cf32d23a7772f172c7dce8a693ffab96166e))
* **deps:** bump dependabot/fetch-metadata from 1.3.0 to 1.3.1 ([fc50474](https://github.com/justintaddei/tcp-ping/commit/fc50474c1a7040296fc9afc1f725342e62a552d1))
* **deps:** bump dependabot/fetch-metadata from 1.3.1 to 1.3.2 ([18f58e1](https://github.com/justintaddei/tcp-ping/commit/18f58e17072867b53f7658bfce70feacf625e2df))
* **deps:** bump dependabot/fetch-metadata from 1.3.2 to 1.4.0 ([380a8aa](https://github.com/justintaddei/tcp-ping/commit/380a8aaa65e0e02889c789afc93145caca25fc5d))
* **deps:** bump ini from 1.3.5 to 1.3.8 ([6632433](https://github.com/justintaddei/tcp-ping/commit/66324338a589b5a28fd20e826e9640b4bbc255c9))
* **deps:** bump minimist from 1.2.5 to 1.2.6 ([e09acbe](https://github.com/justintaddei/tcp-ping/commit/e09acbe7667b4abfc7372bebfd053ed73532c99d))
* **deps:** bump path-parse from 1.0.6 to 1.0.7 ([34bbd84](https://github.com/justintaddei/tcp-ping/commit/34bbd84dd967bea391da87c18f7c11cb5ca107bf))
* **deps:** bump tar from 4.4.13 to 4.4.19 ([e8e5e47](https://github.com/justintaddei/tcp-ping/commit/e8e5e47171eca10146e38225b338e5c7f5c52cbe))
* **deps:** bump tmpl from 1.0.4 to 1.0.5 ([22c6f4c](https://github.com/justintaddei/tcp-ping/commit/22c6f4c9b537286f41f345f3ecd05ab7b5ff4c95))
* update tests to work with jest@29 ([c62b3bc](https://github.com/justintaddei/tcp-ping/commit/c62b3bc64a8ef93d6f8af0938964a009230811fd))

# CHANGE LOG

## v1.2.0

- Remove IP limitation, so that hostnames can be used to probe/ping.


__test__/tcp-ping.test.ts
jest.mock('net')

import { ping, probe } from '../src/index'

describe('Ping localhost', () => {
  test('Ping 127.0.0.1 on an open port', async () => {
    expect.assertions(7)
    await expect(
      ping({ port: 1, attempts: 3 }, (progress, total) => {
        expect(total).toBe(3)
        expect(progress).toBeLessThanOrEqual(total)
      })
    ).resolves.toBeInstanceOf(Object)
  })

  test('Pinging 127.0.0.1 on a closed port times out', async () => {
    await expect(ping({ port: 2, attempts: 3, timeout: 1000 })).resolves.toHaveProperty(
      ['errors', 0, 'error'],
      Error('Request timeout')
    )
  })
})

describe('Probe localhost', () => {
  test('Probe 127.0.0.1 on an open port', async () => {
    await expect(probe(1)).resolves.toBe(true)
  })

  test('Probe 127.0.0.1 on a closed port', async () => {
    await expect(probe(2)).resolves.toBe(false)
  })
})

describe('Valid options', () => {
  // Removed to allow pinging hostnames
  /* test('Throws "Invalid IP" error when an invalid IP is given', async done => {
    await expect(ping({ address: 'not an ip address' })).rejects.toThrowError('Invalid IP')
    await expect(probe(80, 'not an ip address')).rejects.toThrowError('Invalid IP')
    
  }) */

  test('Throws "Negative port" error when the port is less than 1', async () => {
    await expect(ping({ port: 0 })).rejects.toThrowError('Negative port')
    await expect(ping({ port: -100 })).rejects.toThrowError('Negative port')
    await expect(probe(0)).rejects.toThrowError('Negative port')
    await expect(probe(-100)).rejects.toThrowError('Negative port')
  })
})


src/index.ts
import { Socket } from 'net'

/**
 * Ping Options
 */
export interface IPingOptions {
  /**
   * The IP address of the device being pinged
   */
  address: string
  /**
   * The port to ping
   */
  port: number
  /**
   * The number of times
   * `tcp-probe#ping` will attempt to connect to `address`
   */
  attempts: number
  /**
   * The time whichafter each TCP socket will timeout
   */
  timeout: number
}

/**
 * The result of ping attempt that failed
 */
export interface IPingError {
  /**
   * The number of the attempt
   */
  attempt: number
  /**
   * The error that occured
   */
  error: Error
}

/**
 * The results of a ping attempt
 */
export interface IPingResult {
  /**
   * The options that were used for this ping
   */
  options: IPingOptions
  /**
   * The average latency of this ping
   */
  averageLatency: number
  /**
   * The latency of the fastest attempt
   */
  minimumLatency: number
  /**
   * The latency of the slowest attempt
   */
  maximumLatency: number
  /**
   * All the ping attempts that failed
   */
  errors: IPingError[]
}

/**
 * The result of a connection attempt
 */
export interface IConnectionAttempt {
  /**
   * The number of this attempt
   */
  attemptNumber: number
  /**
   * The result of the connection
   */
  result: {
    /**
     * The time it took to connect (e.i. the latency of the connection)
     */
    time?: number
    /**
     * The `Error` that caused the connection to fail
     */
    error?: Error
  }
}

/**
 * Attempts to connect to the given host and returns the `IConnectionResult`
 * @param options The `IPingOptions` to use for this connection
 */
function connect({ address, port, timeout }: IPingOptions): Promise<IConnectionAttempt['result']> {
  return new Promise((resolve) => {
    // Create a new tcp scoket
    const socket = new Socket()
    // Save the current time so we can calculate latency
    const startTime = process.hrtime()

    // Connect to the given host
    socket.connect(port, address, () => {
      // Calculate the latency of the connection
      const [seconds, nanoseconds] = process.hrtime(startTime)

      // Convert the latency from nanoseconds to milliseconds
      // so that the output is easier to work with
      const timeToConnect = (seconds * 1e9 + nanoseconds) / 1e6

      // We don't need the socket anymore
      // so we should destroy it
      socket.destroy()
      // Resolve with the latency of this attempt
      resolve({ time: timeToConnect })
    })

    // Make sure we catch any errors thrown by the socket
    socket.on('error', (error) => {
      // We don't need the socket anymore
      // so we should destroy it
      socket.destroy()
      // Resolve with the error
      resolve({ error })
    })

    // Set the timeout for the connection
    socket.setTimeout(timeout, () => {
      // We don't need the socket anymore
      // so we should destroy it
      socket.destroy()
      // Resolve with a timeout error
      resolve({ error: Error('Request timeout') })
    })
  })
}

/**
 * Pings the given device and report the statistics
 * in the form of an `IPingResult` object
 * @param options The `IPingOptions` object
 */
export async function ping(
  options?: Partial<IPingOptions>,
  progress?: (progress: number, total: number) => void
): Promise<IPingResult> {
  // Default ping options
  const opts: IPingOptions = {
    address: '127.0.0.1',
    attempts: 10,
    port: 80,
    timeout: 3000,
    // Otherwrite default options
    ...options,
  }

  // Removed to allow pinging hostnames
  // Make sure this is a real IP address
  // if (!isIP(opts.address)) throw Error('Invalid IP')

  if (opts.port < 1) throw RangeError('Negative port')

  /**
   * An array of all the connection attempts
   */
  const connectionResults: IConnectionAttempt[] = []

  // Try to connect to the given host
  for (let i = 0; i < opts.attempts; i++) {
    connectionResults.push({
      // i + 1 so the first attempt is `attempt 1`
      // instead of `attempt 0`
      attemptNumber: i + 1,
      result: await connect(opts),
    })

    if (typeof progress === 'function') progress(i + 1, opts.attempts)
  }
  /**
   * The result of this ping
   */
  const result: IPingResult = {
    averageLatency: NaN,
    errors: [],
    maximumLatency: 0,
    minimumLatency: Infinity,
    options: opts,
  }

  /**
   * The sum of the latency of all
   * the successful ping attempts
   */
  let latencySum: number = 0

  // Loop over all the connection results
  for (const attempt of connectionResults) {
    // If `time` is undefined then
    // assume there's an error
    if (typeof attempt.result.time === 'undefined') {
      // Push the error onto the errors array
      result.errors.push({
        attempt: attempt.attemptNumber,
        // If error is undefined then throw an unknown error
        error: attempt.result.error || Error('Unknown error'),
      })
      // We're done with this iteration
      continue
    }

    // Get the latency of this attempt
    const { time } = attempt.result

    // Add it to the sum
    latencySum += time

    // If this attempts latency is less
    // then the current `minimumLatency` then we
    // update `minimumLatency`
    if (time < result.minimumLatency) result.minimumLatency = time

    // If this attempts latency is greater
    // then the current `maximumLatency` then we
    // update `maximumLatency`
    if (time > result.maximumLatency) result.maximumLatency = time
  }

  // Calculate the average latency of all the attempts
  // (excluding the attempts that errored because those
  // didn't return a latency)
  result.averageLatency = latencySum / (connectionResults.length - result.errors.length)

  // Finally, resolve with the result
  return result
}

/**
 * Makes one attempt to reach the host and returns
 * a `boolean` indicating whether or not it was successful.
 * @param port The port to probe
 * @param address The address to probe
 * @param timeout The timeout of the probe
 */
export async function probe(port: number, address: string = '127.0.0.1', timeout: number = 3000): Promise<boolean> {
  // Ping the host
  const result = await ping({ address, port, timeout, attempts: 1 })

  // If there aren't any error then the device is reachable
  return result.errors.length === 0
}
